### PR TITLE
chore: visual refresh on buttons and anchors

### PR DESCRIPTION
**Pull request by khamudom**
# Pull Request

## 📖 Description

The default height has been updated to 32px and the corner radius is 6px. The focus border has also been updated to 3px.

Before and After
The order down is rest, hover, active, focus, and disabled states.

Default/Neutral
image

Accent
image

Lightweight
image

Outline
image

Stealth
image

The visual for Anchor is the same for neutral, accent, lightweight, outline, and stealth, except Anchor, does not have a disabled state.

Before and After for Anchor Hypertext
image

 
### General

- [ ] I have included a change request file using `$ yarn change`
- [ ] I have added tests for my changes.
- [x] I have tested my changes.
- [ ] I have updated the project documentation to reflect my changes.
- [x] I have read the [CONTRIBUTING](https://github.com/Microsoft/fast/blob/master/CONTRIBUTING.md) documentation and followed the [standards](https://www.fast.design/docs/community/code-of-conduct/#our-standards) for this project.

### Component-specific

- [ ] I have added a new component
- [ ] I have modified an existing component
- [ ] I have updated the [definition file](https://github.com/Microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#definition)
- [ ] I have updated the [configuration file](https://github.com/Microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#configuration)

### CODE REVIEW

**Comment by janechu** [comment]
Should the focus outline (or border) for the anchor be that tight?

**Comment by khamudom** [comment]
> Should the focus outline (or border) for the anchor be that tight?

Yeah, I personally think it is too tight. I kept it as is because there is no design for hypertext. I think we can just keep focus with an underline, which is the current visual. @bheston any suggestion? But for now, I will update to an underline, unless there are any objections or other ideas.

**Comment by khamudom** [comment]
hypertext updated
image

File: packages/web-components/fast-components/src/styles/patterns/button.ts [hunk 1]
```diff
@@ -38,457 +45,418 @@ import {
...
+                  color: ${SystemColors.HighlightText};
+              }
+              :host .control:${focusVisible} {
+                  forced-color-adjust: none;
+                  border-color: ${SystemColors.ButtonText} !important;
+                  box-shadow: 0 0 0 calc(${focusStrokeWidth} * 1px) ${SystemColors.ButtonText} !important;
+                  background: ${SystemColors.ButtonFace};
+                  color: ${SystemColors.ButtonText};
+              }
+              :host([href]) .control {
```
Comment (robarbms) [inline packages/web-components/fast-components/src/styles/patterns/button.ts:146]: I couldn't find href used in button.ts or in button.template.ts. Are these shared styles or should all of the :host([href]) styles be in anchor.styles.ts? I get that you didn't put them here, just a broader question. I'll mark this comment as resolved.

File: packages/web-components/fast-components/src/styles/patterns/button.ts [hunk 2]
```diff
@@ -38,457 +45,418 @@ import {
...
-              fill: currentColor;
-            }
-        `
-    )
-);
+export const BaseButtonStyles = (
+    context: ElementDefinitionContext,
+    definition: FoundationElementDefinition,
+    interactivitySelector: string = ""
+) =>
```
Comment (robarbms) [inline packages/web-components/fast-components/src/styles/patterns/button.ts:52]: You are calling out the opposite of the interactivitySelector explicitly in buton.styles.ts and anchor.styles.ts above. Would it make sense to create that here and move the styles into BaseButtonStyles?

```
const inactiveSelector = interactivitySelector.indexOf(":not") >= 0 ? 
    interactivitySelector.replace(/(.*):not\(([^)]*)\)(.*)/, '$1$2$3') : 
    `:not(${interactivitySelector})`; 
```

**Review by robarbms — COMMENTED** [review]
There are are a lot of !important statements included. Is there anyway to remove them. As soon as you add 1, you need to add a hundred. It turns into an arms race for CSS selectors and becomes really hard to manage. Another drawback to !important is that it becomes impossible to override it from inline. Inline styles will supersede stylesheet styles, but for !important statements, the opposite is true. Which means that a `style="color: #000!important"` will not override a `:host { color: #fff!important; }` in a stylesheet.

**Comment by janechu** [comment]
Who's approvals should we consider as necessary (blocking) for these PRs? @bheston? I just want to make sure if I sign off on this, I am not the "authority" on it as I'm not aware of all of the style decisions that have been made.

**Comment by khamudom** [comment]
> Who's approvals should we consider as necessary (blocking) for these PRs? @bheston? I just want to make sure if I sign off on this, I am not the "authority" on it as I'm not aware of all of the style decisions that have been made.

As far as the visual approval, with an exception of a few components like Picker and Calendar, before Jason Reynolds left, we went through all the components and he approved the visual changes. Then I created the PR that was up for review (https://github.com/microsoft/fast/pull/5167) 
Now that we are breaking up the changes into smaller PRs and adding them to a feature branch, maybe the approval is more about the code and not the visual. If we still accept Jason Reynolds as the person who signs off on the visual refresh since that has not changed in these PRs. Not to say, we can't give feedback on the visual, because that is also welcomed. For example, the focus on the hypertext button. That was good feedback. Just my thoughts.

**Comment by EisenbergEffect** [comment]
I think @bheston and @chrisdholt should probably take a look at and be the ones to approve the visual refresh PRs.

File: packages/web-components/fast-components/src/button/button.styles.ts [hunk 3]
```diff
@@ -31,129 +27,58 @@ export const buttonStyles: (
...
-                    color: ${SystemColors.GrayText};
-                    cursor: ${disabledCursor};
+                :host([disabled]) .control {
                     opacity: 1;
+                    background: ${SystemColors.ButtonFace};
+                    border-color: ${SystemColors.GrayText} !important;
+                    box-shadow: none !important;
+                    color: ${SystemColors.GrayText};
+                    fill: currentcolor;
                 }
```
[the remark this answers is not in the document]
Comment (khamudom) [reply packages/web-components/fast-components/src/button/button.styles.ts:46]: using `!important` is not necessary, I will remove them

File: packages/web-components/fast-components/src/styles/patterns/button.ts (hunk 1 above)
Comment (khamudom) [reply packages/web-components/fast-components/src/styles/patterns/button.ts:146]: I think refactoring to move, the selectors with `href`, into anchor style, is a good idea.

**Review by janechu — APPROVED** [review]

**Comment by bheston** [comment]
> > Should the focus outline (or border) for the anchor be that tight?
> 
> Yeah, I personally think it is too tight. I kept it as is because there is no design for hypertext. I think we can just keep focus with an underline, which is the current visual. @bheston any suggestion? But for now, I will update to an underline, unless there are any objections or other ideas.

I would really love for text underline to have a setting for how far away from the text it should be in addition to the one that specifies the thickness. Lacking that, I still think it's preferable to faking an underline with a border because the underline is cut for descenders. There's no great way to do that with a border or other treatment.

I'd also really love for comments to be conversations in GitHub just like if they were based on a line of code. Two things to hope for.

**Review by chrisdholt — APPROVED** [review]

File: packages/web-components/fast-components/src/styles/patterns/button.ts (hunk 2 above)
Comment (khamudom) [reply packages/web-components/fast-components/src/styles/patterns/button.ts:53]: updated. thanks

File: packages/web-components/fast-components/src/anchor/anchor.styles.ts [hunk 4]
```diff
@@ -19,11 +29,173 @@ export const anchorStyles: (
...
+            css`
+              :host([href]) .control {
+                  border-color: ${SystemColors.LinkText};
+                  background: ${SystemColors.ButtonFace};
+                  color: ${SystemColors.LinkText};
+                  fill: currentcolor;
+              }
+              :host([href]) .control:hover,
+              :host(.neutral[href]) .control:hover,
+              :host(.outline[href]) .control:hover {
```
Comment (bheston) [inline packages/web-components/fast-components/src/anchor/anchor.styles.ts:52]: The `.outline` styles should move down to the outline appearance behavior.

**Review by bheston — APPROVED** [review]

**Pull request merged**